sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala

```diff
@@ -157,19 +157,45 @@ abstract class AggregationIterator(
       inputAttributes: Seq[Attribute]): (InternalRow, InternalRow) => Unit = {
     val joinedRow = new JoinedRow
     if (expressions.nonEmpty) {
-      val mergeExpressions = functions.zip(expressions).flatMap {
-        case (ae: DeclarativeAggregate, expression) =>
-          expression.mode match {
-            case Partial | Complete => ae.updateExpressions
-            case PartialMerge | Final => ae.mergeExpressions
+      val mergeExpressions =
+        functions.zip(expressions.map(ae => (ae.mode, ae.isDistinct, ae.filter))).flatMap {
+          case (ae: DeclarativeAggregate, (mode, isDistinct, filter)) =>
+            mode match {
+              case Partial | Complete =>
+                if (filter.isDefined) {
+                  ae.updateExpressions.zip(ae.aggBufferAttributes).map {
+                    case (updateExpr, attr) => If(filter.get, updateExpr, attr)
+                  }
+                } else {
+                  ae.updateExpressions
+                }
+              case PartialMerge | Final => ae.mergeExpressions
+            }
+          case (agg: AggregateFunction, _) => Seq.fill(agg.aggBufferAttributes.length)(NoOp)
+        }
+      // Initialize predicates for aggregate functions if necessary
+      val predicateOptions = expressions.map {
+        case AggregateExpression(_, mode, _, Some(filter), _) =>
+          mode match {
+            case Partial | Complete =>
+              val filterAttrs = filter.references.toSeq
+              val predicate = Predicate.create(filter, inputAttributes ++ filterAttrs)
```

Comment:
why do we need `inputAttributes` here?

Reply:
The attributes reference by filter is also needed.

Reply:
what we are doing here is binding the attributes. For example, the input schema is `[a, b, c]`, and the filter is `b > 1`, then after binding it's `col(1) > 1`.

That said, the attributes we provide here must match the input row, otherwise we may read the wrong column at runtime.

I don't see we pull out the filter reference columns and join them with the original input row.

Reply:
I have taken a try and occur issue as below : 
```
org.apache.spark.sql.catalyst.errors.package$TreeNodeException
Binding attribute, tree: null
```

Reply:
Then we must figure out why. My point is, the attributes that used to bind expressions, must match the schema of the input row.

I don't see how can `inputAttributes ++ filterAttrs` match the input row.

Reply:
I find out the reason 
The filter expression not convert to attributes here!
https://github.com/apache/spark/blob/91fe90e79a2a07153eb89921dbb7528490f44b64/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala#L241
The child plan of first aggretate is `Expand` and it's projection is converted to attributes already.
The operator will used for first aggregate.
So the filter expression can't apply in first aggregate.